tech-specs/src/explore/design/ibc.md

```diff
@@ -0,0 +1,127 @@
+
+The transaction is given an IBC packet or message which specifies what to do. Because the packet or message has been encoded and stored in transaction data and there are some types of messages, it has to check what packet or message is given first and then decodes the message or packet. After the process according to the packet or message, it emits an IBC event.
+
+- Transaction given a message
+  ```rust
+  #[transaction]
+  fn apply_tx(tx_data: Vec<u8>) {
+      let signed =
+          key::ed25519::SignedTxData::try_from_slice(&tx_data[..]).unwrap();
+      let messages: Vec<Any> = prost::Message::decode(&signed.data[..]).unwrap();
+      let result = match message.type_url.as_str() {
+          // decode each message: refer to `ibc-rs`
+          create_client::TYPE_URL => {
+              let domain_msg = create_client::MsgCreateAnyClient::decode_vec(&any_msg.value)
+                  .map_err(|e| Kind::MalformedMessageBytes.context(e))?;
+              ibc_create_client(domain_msg)
+          }
+          // other messages
+          ...
+      };
+      match &result {
+          Ok(output) => emit_event(output.events),
+          Err(e) => {
+              tx::log_string(format!("IBC operation faild {}", e));
+              unreachable!()
+          }
+      }
+  }
+  ```
+
+- Transaction given a packet
+  ```rust
+  #[transaction]
+  fn apply_tx(tx_data: Vec<u8>) {
+      let signed =
+          key::ed25519::SignedTxData::try_from_slice(&tx_data[..]).unwrap();
+      let packet: Packet = prost::Message::decode(&signed.data[..]).unwrap();
+      let data: FungibleTokenPacketData = prost::Message::decode(&packet.data[..]).unwrap();
+      let result = ibc_transfer(data);
+      match &result {
+          Ok(output) => emit_event(output.events),
+          Err(e) => {
+              tx::log_string(format!("IBC transfer faild {}", e));
+              unreachable!()
+          }
+      }
+  }
+  ```
+
+### Make a packet
+IBC-related transaction for some messages or packets makes a packet. For example, when a transaction wants to transfer a token between ledgers, it should make a packet including `FungibleTokenPacketData` to specify the sender, receiver, token, and amount.
+
+### Store a proof and state
+A proof can be stored to the storage on the ledger. Its storage key should be prefixed (e.g. `ibc/`) to protect them from other storage operations. The paths(keys) for Tendermint client are defined by [ICS 24](https://github.com/cosmos/ibc/blob/master/spec/core/ics-024-host-requirements/README.md#path-space).
```

Comment:
We don't need to store proofs, we just need to validate them.

Reply:
Thanks, I misunderstood that. I've modified that.